[#41821] Non working days & duration in scheduling

cbliard opened the pull request:
Related to: Non working days & duration in scheduling https://community.openproject.org/work_packages/41821

cbliard committed:
[#41821] Can compute duration based on defined non working days

Related to: Non working days & duration in scheduling https://community.openproject.org/work_packages/41821

cbliard committed:
[#41821] Set duration when from date or to date is nil

cbliard committed:
Use Days service to compute duration

cbliard committed:
Can assert work packages schedule through visual representation

Helper `let_schedule` defines work packages, their dates, and their relations through a visual chart.

Helper `expect_schedule` runs expectations of work packages against a schedule.

cbliard committed:
[#41821] Partially schedule according to non-working days

cbliard committed:
Reschedule correctly if delta covers non-working days

cbliard committed:
Fix test

cbliard committed:
Can declare schedule dependencies in any order

cbliard committed:
[#41821] Respect non-working days in soonest start date

cbliard committed:
Schedule helpers can display a self representation

The self representation is normalized and only keeps the origin Monday
as reference. There are no properties either.

By adding the ability to add real work packages, it becomes possible to
compare two representation during tests. That's exactly what the
`match_schedule` rspec matcher is doing. It compares the visual
representation instead of comparing date attributes. It should make it
easier to spot differences in scheduling shapes.

cbliard committed:
Add more scheduling test cases

So far, the code still works without any additions.

cbliard committed:
Remove dead code

cbliard committed:
[#41821] Calculate due date from start date and duration

cbliard committed:
Split ScheduleHelpers module in smaller files

cbliard committed:
[#41821] Reschedule on new follows relation or parent

* `ChartBuilder` supports 'child of <wp>' property

* `match_schedule` normalizes charts to get same size and order before
  comparison

cbliard committed:
[#41821] Add test cases for single successor having a parent

cbliard committed:
Use better method name

It has nothing to do with followers. It just creates a child.

cbliard committed:
Fix duration calculation in work package factory

cbliard committed:
[#41821] Add test cases for single successor having two children

cbliard committed:
[#43142] Keep the due date when rescheduling if possible

cbliard committed:
Ensure due_date is after start_date when rescheduling by parent

When setting parent of a work package without start date, it will set
the start date to be the parent soonest start date. It also has to
ensure that the end date is changed if it was set to a date earlier than
the new start date.

cbliard committed:
[#41821] Set dates during working days when setting parent

Also:
* ScheduleHelpers::Chart can display work packages which ignore
  non-working days
* ScheduleHelpers::ChartBuilder understands "duration <int>" property

cbliard commented:
@dombesz The PR is ready for review. Thanks for taking over it.

dombesz commented on app/services/work_packages/shared/working_days.rb:
```diff
@@ -0,0 +1,121 @@
+        WeekDay.pluck(:day, :working).each do |day, working|
+          @working_week_days[day] = working
+        end
+        @working_week_days[0] = @working_week_days[7] # Sunday is 7 in iso or 0 in other implementations
```
The `working?(date)` method seems to have no tests and this line doesn't seems to be correct.
```
pry> @working_week_days
=> [true, true, true, true, true, true, false, false]
pry> @working_week_days[0] = @working_week_days[7]
=> false
pry > @working_week_days
=> [false, true, true, true, true, true, false, false]
```

Should we just rather rotate the array elements?

```
pry> @working_week_days
=> [true, true, true, true, true, true, false, false]
@working_week_days.rotate!(7)
=> [false, true, true, true, true, true, true, false]
```
Or:
```
pry> @working_week_days
=> [true, true, true, true, true, true, false, false]
@working_week_days.prepend(@working_week_days.pop)
=> [false, true, true, true, true, true, true, false]
```

cbliard replied:
It builds an array with information about which day is working or not. Because Sunday is sometimes on 0 index (because that's what Date#wday returns), and sometimes on 7 index (because we chose to conform to ISO), an array of size 8 is built with the information on index 0 and 7 being duplicated. So the array is [Sunday, Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday]

Because database table for week days could be empty, it builds it with `true` values everywhere first.
```
        @working_week_days = [true] * 8
```
Then picks the values from the db which will fill index 1-7
```
        WeekDay.pluck(:day, :working).each do |day, working|
          @working_week_days[day] = working
        end
```
And finally copies the value for Sunday at index 7 to the index 0
```
         @working_week_days[0] = @working_week_days[7] # Sunday is 7 in iso or 0 in other implementations
```

So with Saturday and Sunday being weekend days, the array should be 
```
[false, true, true, true, true, true, false, false]
```

Is is clearer now? I'll add some comments next to the code.

cbliard replied:
comments added in [9568abf](https://github.com/opf/openproject/pull/10851/commits/9568abf2e279f24ae6cd8fc147f39f50e46a8e85)

dombesz commented on app/services/work_packages/set_schedule_service.rb:
```diff
@@ -56,11 +56,23 @@ def call(changed_attributes = %i(start_date due_date))
       .select { |wp| wp.start_date.nil? && wp.parent }
-      .each { |wp| wp.start_date = wp.parent.soonest_start }
+      .each do |wp|
+        days = WorkPackages::Shared::Days.for(wp)
```
Could this get slow in case of a large number of wp's to reschedule?

cbliard replied:
Maybe, but maybe not as there are SQL caches. Let's keep it simple first, and improve its performance if we observe a performance issue.

dombesz commented on app/services/work_packages/set_schedule_service.rb:
```diff
@@ -137,45 +149,60 @@ def reschedule_by_predecessors(scheduled, dependency)
+    min_start_delta = days.delta(previous: scheduled.start_date || min_start_date, current: min_start_date)
+    required_delta = [min_start_delta, [moved_delta, 0].min].max
+
+    new_start_date = WorkPackages::Shared::Days.for(scheduled).add_days(scheduled.start_date, required_delta)
```
We should rather instantiate a days object for the scheduled to benefit from the caching.

```ruby
scheduled_days = WorkPackages::Shared::Days.for(scheduled)
```

cbliard replied:
yep, we can do that.

cbliard replied:
added in [9568abf](https://github.com/opf/openproject/pull/10851/commits/9568abf2e279f24ae6cd8fc147f39f50e46a8e85)

dombesz commented on app/services/work_packages/shared/days.rb:
```diff
@@ -0,0 +1,40 @@
+    class Days
+      # Returns the right day computation instance for the given instance.
+      def self.for(work_package)
+        return AllDays.new unless OpenProject::FeatureDecisions.work_packages_duration_field_active?
```
Perhaps it would make sense to cache the `AllDays` and `WorkingDays`, because those instances would be the same for all the work packages. This way we can avoid querying the database for working days for each work package.

cbliard replied:
Yep, I thought about that, and had the issue of "when should the cache be invalidated?"

So I'd like to measure the performance impact first, and if performance is ok, it will save us headaches about invalidating caches and keep our code simple.

dombesz commented on spec/services/work_packages/set_schedule_service_working_days_spec.rb:
```diff
@@ -0,0 +1,1200 @@
+        let_schedule(<<~CHART, ignore_non_working_days: false)
+          days          | MTWTFSS   |
+          work_package  | X         |
+          follower      |     X..XX  | follows work_package
```
There is an extra space here.

cbliard replied:
Nice spot ;)

dombesz commented on spec/services/work_packages/set_schedule_service_working_days_spec.rb:
```diff
@@ -0,0 +1,1200 @@
+          expect_schedule(subject.all_results, <<~CHART)
+            days          | mtwtfssmtwtfssMTWTFSS |
+            work_package  |  X                    |
+            follower      |            X..X       |
```
```suggestion
            follower      |            X..X       |
            annoyer       |    XX..XX             |
```
Since we are also testing that the follower has a 2 days distance from the `annoyer`, I would add the `annoyer` to the expected chart for better visual cueing.

cbliard replied:
If we do that, then the test will fail because `annoyer` was not present in `subject.all_results`. I agree it would be better to have it for better visual cueing, but I did not find a good way to do so. Ideas? 

My current idea: maybe a `#` before the line to ignore it? But then it would not be tested. What if the line gets out of date?

dombesz commented on spec/services/work_packages/set_schedule_service_working_days_spec.rb:
```diff
@@ -0,0 +1,1200 @@
+          expect(subject.all_results).to match_schedule(<<~CHART)
+            days          | mtwtfssmtwtfssMTWTFSS |
+            work_package  |  X                    |
+            follower      |            X..XXX     |
```
Same as above.
```suggestion
            follower      |            X..XXX     |
            annoyer       |    XX..XX             |
```

cbliard replied:
yep same. I'm resolving to only have one conversation about it.

dombesz commented on spec/services/work_packages/set_schedule_service_working_days_spec.rb:
```diff
@@ -0,0 +1,1200 @@
+      end
+    end
+
+    context 'when moving backwards with the parent having a predecessor not limiting movement' do
```
?
```suggestion
    context 'when moving backwards with the parent having a predecessor limiting movement' do
```

cbliard replied:
yep, nice catch

cbliard replied:
fixed in [9568abf](https://github.com/opf/openproject/pull/10851/commits/9568abf2e279f24ae6cd8fc147f39f50e46a8e85)

dombesz commented on spec/services/work_packages/set_schedule_service_working_days_spec.rb:
```diff
@@ -0,0 +1,1200 @@
+        CHART
+      end
+    end
+
```
For the sake of completeness, I would add this usecase as well.
```suggestion

    context 'when moving forward with some delay between the followers' do
      let_schedule(<<~CHART, ignore_non_working_days: false)
        days         | MTWTFSSm     sm     sm     |
        work_package | ]                          |
        follower1    |  XXX                       | follows work_package
        follower2    |        XXXX                | follows follower1 with delay 3
        follower3    |                 XXX..XX    | follows follower2
        follower4    |                         XX | follows follower3
      CHART

      before do
        change_schedule([work_package], <<~CHART)
          days         | MTWTFSS |
          work_package |    ]    |
        CHART
      end

      it 'reschedules all the followers keeping the delay for follower2 and compacting the rest' do
        expect(subject.all_results).to match_schedule(<<~CHART)
          days         | MTWTFSSm     sm     sm     sm |
          work_package |    ]                          |
          follower1    |     X..XX                     |
          follower2    |               XXXX            |
          follower3    |                   X..XXXX     |
          follower4    |                          X..X |
        CHART
      end
    end

```

cbliard replied:
yep!

cbliard replied:
added in [9568abf](https://github.com/opf/openproject/pull/10851/commits/9568abf2e279f24ae6cd8fc147f39f50e46a8e85)

dombesz reviewed (changes requested):
Excellent job @cbliard, I really like the visual charts as setup and expectation too! 👏🎉
I have a few observations, but nothing serious.

cbliard committed:
Add comments and add additional test

Based on pull requests comments from Attila.

Co-authored-by: Dombi Attila <dombesz.attila@gmail.com>

cbliard committed:
Fix test

The test was failing between Monday and Thursday because the end date was before the start date, and thus the displayed duration was 0 days

cbliard committed:
Fix test again

Expected value was wrong

cbliard committed:
Merge branch 'dev' into implementation/41821-non-working-days-and-duration-in-scheduling

cbliard committed:
Fix unit test

as duration was 1, and start date was available, due date was computed. Making duration nil by default fixed it.
